### ISSUE TITLE
Functionality and Installation questions for macOS and Ubuntu

### ISSUE THREAD

**Issue by RielFox** [M-J-Murray/MAMEToolkit#7]
Hey there,

Some months ago I discovered MAME's Lua engine and I started working on a reinforcement learning project that would train on all four currently emulated Tekken games (1,2,3,Tag). I wrote a Lua script that would control the emulator, selecting and loading randomly one of the four tekken games and wait for a demo cycle. The idea was that it would communiciate with a python agent every timestep via sockets, passing state information such as p1_hp, p2_hp, p1_score, p2_score, remaining_time (which would be read from memory values) as well as a frame at the current timestep, and would receive an action to perform. I stumbled on some limitations like how sockets might be too slow and also the Lua engine of MAME cannot take extensions for portability reasons. As such, I am thinking of migrating my project to this python framework. I have a few questions.

The first questions is, is this framework as powerful as Mame's Lua library? Will I be able to change games on the fly without human interaction? My Lua script also relied extensively on emu.wait() calls, which I'm not sure if direct emu commands can be given. But this framework seems to bypass a few problems, like the game warning screen, where the only way to be skipped was with a -str 299 argument that gave you max running time of 299 seconds.

Secondly, I tried installing it to my macOS installation and I will try with Ubuntu as well.
I went into my Anaconda environment with my ML tools and used "pip install MAMEToolkit".
When I tried to run it in python 

>from MAMEToolkit.emulator import see_games
>see_games()

I got "cannot execute binary file
/bin/sh: /Users/rielfox/anaconda3/envs/tensorflow/lib/python3.6/site-packages/MAMEToolkit/emulator/mame/mame: Undefined error: 0". I assume you have to put a compiled version of MAME in the folder   "/MAMEToolkit/emulator/mame". I have a compiled version of macOS Mame, but I believe it wants the compiled modified version of MAME found here "https://github.com/M-J-Murray/mame". I tried downloading and using "make" with terminal at that folder but I get

`GCC 4.2.1 detected
Linking genie
ld: library not found for -lgcc_s.10.4
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[1]: *** [../../bin/darwin/genie] Error 1
make: *** [3rdparty/genie/bin/darwin/genie] Error 2`

I'm not sure if it will only work on GNU/Linux.

Thanks!

**Comment by M-J-Murray** [comment]
Hello

This library works by interacting with a modified version of mame. It uses the lua engine to send frame data to the library via linux fifo pipes. These pipes also allow for the application to be efficiently stepped along. I also experimented with emu.wait() calls, but fifo pipes were the most reliable and efficient. 
You will be able to stop emulations dynamically and reopen them with different games. Meaning you can automate this in your python script.
This library was developed on Ubuntu, but it should work for all Linux systems. The reason the library doesn't work on other platforms is because I am using the Linux fifo pipes to communicate with the emulator. At this time I am not aware of similar mechanisms on other operating systems.

Michael

**Comment by M-J-Murray** [comment]
You do not need to do anything special with mame to use this library. All you need to do is get a python environment, I recommend you use virtual environment or anoconda. Then go to your pip file inside your python environment. And just run './pip install MAMEToolkit' and you should be ready to go.

**Comment by M-J-Murray** [comment]
The pip repo comes with the compiled modified version of mame. The reason it doesn't exist in the github repo is because github has a 50mb file limit on free repos. Let me know if you have any more trouble.

**Comment by RielFox** [comment]
Hey,
I tried some code to see that everything works in my conda environment.
'
import numpy as np
import tensorflow as tf
from MAMEToolkit.emulator import see_games

see_games()
'
I get "./mame: error while loading shared libraries: libQt5Widgets.so.5: cannot open shared object file: No such file or directory". It ends up I have all the Qt files needed in my installation, and I even re downloaded Qt and installed for Ubuntu.

I am trying to follow these steps now: http://doc.qt.io/qt-5/linux-deployment.html
Will update if I get it going.

Thanks!

**Comment by chayinfeng** [comment]
I solved depency problems like libqt5widgets5 by sudo apt-get install. While when I run the simulator, I got 
AttributeError: 'Emulator' object has no attribute 'actionPipe'.

**Comment by RielFox** [comment]
My previous Ubuntu installation proved to be unstable and ended up in Kernel Panics and frequent crashes after a while. In a new Ubuntu and lubuntu installation, inside my conda environment, I did:
>sudo apt-get install libsdl2-dev
>sudo apt-get install libsdl2-ttf-dev
>sudo apt-get install libqt5widgets5
and it worked, I was able to start MAME from Python with 'see_games()'.

I'm not sure why it wasn't working before, as I tried everything listed, but I assume my failing system/ssd might have caused problems.
